### PR TITLE
Eth2-to-Near-relay: add DAO contract interface

### PR DESCRIPTION
**Pull request by karim-en**
Implement a wrapper for Sputnik DAO contract.
* Added the needed DAO APIs 
* Minor code improvements and apply fmt

More details about the DAO are here 
https://github.com/near-daos/sputnik-dao-contract

### CODE REVIEW

File: eth2near/contract_wrapper/src/dao_contract.rs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    Approve = 0x0,
+    Reject = 0x1,
+    Remove = 0x2,
+}
+
+/// Proposal that are sent to this DAO.
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Debug))]
+#[serde(crate = "near_sdk::serde")]
+pub struct Proposal {
```
Comment (olga24912) [inline eth2near/contract_wrapper/src/dao_contract.rs:86]: I think It is not exactly `Proposal`. `ProposalInput` looks more like Proposal. It is more looks like Proposal with a lot of supporting information. May be `ProposalState` is better?

File: eth2near/contract_wrapper/src/dao_contract.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    pub fn get_proposal(&self, id: u64) -> Result<ProposalOutput, Box<dyn Error>> {
+        let response = self.contract_wrapper.call_view_function(
+            "get_proposal".to_string(),
+            json!({ "id": id }).to_string().into_bytes(),
+        )?;
+
+        Ok(serde_json::from_slice(response.as_slice())?)
+    }
+
+    pub fn add_proposal(&mut self, proposal: ProposalInput) -> Result<u64, Box<dyn Error>> {
```
Comment (olga24912) [inline eth2near/contract_wrapper/src/dao_contract.rs:166]: Is this function should be public? Or it is going to be use only inside submit_light_client_update_proposale?

File: eth2near/contract_wrapper/src/dao_contract.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+            serde_json::to_string(&proposal).unwrap().into_bytes(),
+            Some(1 * near_sdk::ONE_NEAR),
+            None,
+        )?;
+
+        self.last_proposal_id = serde_json::from_slice(response.as_slice())?;
+        Ok(self.last_proposal_id)
+    }
+
+    pub fn act_proposal(&self, id: u64, action: Action) -> Result<(), Box<dyn Error>> {
```
Comment (olga24912) [inline eth2near/contract_wrapper/src/dao_contract.rs:178]: What this function doing and what the difference between add proposal and act proposal?

**Review by olga24912 — COMMENTED** [review]
I don't understand this code well. I think I need more theory about how SputnikDAO in this case works. 

Also we should create trait for eth_client_contract and create eth_client_contract_with_dao to encapsulate all work with DAO. And I don't exactly understand how exactly we should organize communication with DAO and with checking proposal status. And also I start thinking about deleting contract_wrapper_trait and stay only with one near_contract_wrapper.

File: eth2near/contract_wrapper/src/dao_contract.rs (hunk 1 above)
Comment (karim-en) [reply eth2near/contract_wrapper/src/dao_contract.rs:86]: All types are migrated from here, so let's keep the same naming 
https://github.com/near-daos/sputnik-dao-contract/blob/main/sputnikdao2/src/proposals.rs

File: eth2near/contract_wrapper/src/dao_contract.rs (hunk 2 above)
Comment (karim-en) [reply eth2near/contract_wrapper/src/dao_contract.rs:166]: This is not an internal method, so it should be public even if only the helper method is used now.

File: eth2near/contract_wrapper/src/dao_contract.rs (hunk 3 above)
Comment (karim-en) [reply eth2near/contract_wrapper/src/dao_contract.rs:178]: Check this  https://github.com/near-daos/sputnik-dao-contract#approve-proposal

File: eth2near/contract_wrapper/src/dao_contract.rs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    /// Kind of proposal with relevant information.
+    pub kind: ProposalKind,
+    /// Current status of the proposal.
+    pub status: ProposalStatus,
+    /// Count of votes per role per decision: yes / no / spam.
+    pub vote_counts: HashMap<String, [Balance; 3]>,
+    /// Map of who voted and how.
+    pub votes: HashMap<AccountId, Vote>,
+    /// Submission time (for voting period).
+    pub submission_time: U64,
```
[the remark this answers is not in the document]
Comment (karim-en) [reply eth2near/contract_wrapper/src/dao_contract.rs:100]: Check this https://github.com/near-daos/sputnik-dao-contract/blob/main/sputnikdao2/src/proposals.rs#L246

File: eth2near/contract_wrapper/src/eth_client_contract.rs [hunk 5]
```diff
@@ -60,22 +85,47 @@ impl EthClientContract {
...
+            .call_change_method(
+                "register_submitter".to_string(),
+                json!({}).to_string().into_bytes(),
+                Some(self.get_min_deposit()?),
+                None,
+            )
+            .map(|_| {})
+    }
+
+    pub async fn get_light_client_state(&self) -> Result<LightClientState, Box<dyn Error>> {
```
[the remark this answers is not in the document]
Comment (karim-en) [reply eth2near/contract_wrapper/src/eth_client_contract.rs:113]: The async was added by mistake when I moved my async implementation to yours

[the remark this answers is not in the document]
Comment (karim-en) [reply eth2near/contract_wrapper/src/eth_client_contract.rs]: Fixed dee4c8320c91a6d7b6cddc187511878c5fd7161f

File: eth2near/contract_wrapper/src/dao_contract.rs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+impl DAOContract {
+    pub fn new(contract_wrapper: Box<dyn ContractWrapper>) -> Self {
+        DAOContract {
+            last_proposal_id: 0,
+            contract_wrapper,
+        }
+    }
+
+    pub fn get_last_added_proposal(&self) -> u64 {
```
[the remark this answers is not in the document]
Comment (karim-en) [reply eth2near/contract_wrapper/src/dao_contract.rs]: Removed dee4c8320c91a6d7b6cddc187511878c5fd7161f

File: eth2near/contract_wrapper/src/dao_contract.rs [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        return self.last_proposal_id;
+    }
+
+    pub fn get_last_proposal_id(&self) -> Result<u64, Box<dyn Error>> {
```
[the remark this answers is not in the document]
Comment (karim-en) [reply eth2near/contract_wrapper/src/dao_contract.rs:143]: Removed dee4c8320c91a6d7b6cddc187511878c5fd7161f

File: eth2near/contract_wrapper/src/contract_wrapper_trait.rs [hunk 8]
```diff
@@ -1,6 +1,25 @@
...
+        &self,
+        method_name: String,
+        args: Vec<u8>,
+    ) -> Result<Vec<u8>, Box<dyn std::error::Error>>;
+
+    fn call_change_method_batch(
+        &self,
+        method_name: Vec<String>,
+        args: Vec<Vec<u8>>,
+        deposit: Vec<Balance>,
```
[the remark this answers is not in the document]
Comment (karim-en) [reply eth2near/contract_wrapper/src/contract_wrapper_trait.rs]: Fixed 23e564b1d96681315b421ffb7a10ddca2a1bf132

File: eth2near/contract_wrapper/src/dao_contract.rs [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+    ) -> Result<u64, Box<dyn Error>> {
+        let raw_update = update.try_to_vec().unwrap();
+        let update_hash = near_primitives::hash::hash(&raw_update);
+        let args = Base64VecU8::from(raw_update);
+
+        let action = ActionCall {
+            method_name: "submit_beacon_chain_light_client_update".to_string(),
+            args: Base64VecU8::from(args.try_to_vec().unwrap()),
+            deposit: (1 * near_sdk::ONE_NEAR).into(),
+            gas: (290 * Gas::ONE_TERA.0).into(),
```
Comment (sept-en) [inline eth2near/contract_wrapper/src/dao_contract.rs:200]: Why the value is 290? Make it a constant, please.

**Review by sept-en — CHANGES_REQUESTED** [review]

**Review by sept-en — APPROVED** [review]

**Review by sept-en — APPROVED** [review]

**Pull request merged**